Migrate from glibc's libcrypt to libxcrypt

mweinelt opened the pull request:
This PR hopefully paves the way for a libxcrypt migration.

For that we disable crypt suppor in glibc and find out what packages break.
On these packages we add libxcrypt and things should build again.

Ideally we can achieve a shorter path to libxcrypt than drawn up in #112371.

mweinelt committed:
glibc: make crypt support optional

The libcrypt library is going to be replaced with libxcrypt in packages
that require it.

vcunat commented:
https://hydra.nixos.org/jobset/nixos/pr-181764-libxcrypt

mweinelt committed:
libxcrypt: use tarball src, prune dependencies

vcunat commented:
(gcc still doesn't build for me)

mweinelt commented:
Yep, I was taking a look at perl yesterday, since libxcrypt requires it and therefore causes an infinite recursion. Two ideas came up:

- Build Perl once without crypt support
- Build Perl once with glibc's libcrypt

Went for the second approach this morning, but I currently can't get perl to find `glibc.libcrypt`.

```
❯ nix-build -A perl
error: attribute 'libcrypt' missing

       at /home/hexa/git/nixos/libxcrypt/pkgs/top-level/all-packages.nix:19795:40:

        19794|   libxcrypt = callPackage ../development/libraries/libxcrypt {
        19795|     perl = perl.override { libxcrypt = glibc.libcrypt; };
             |                                        ^
        19796|   };
```

I can build `glibc.libcrypt` otherwise.

mweinelt committed:
perl: allow building without libcrypt

mweinelt committed:
libxcrypt: build using perl without libcrypt support

fpletz committed:
gcc11: fix build with libxcrypt

fpletz committed:
cpython: fix build with libxcrypt

fpletz committed:
cyrus_sasl: fix build with libxcrypt

fpletz committed:
libxslt: fix build with libxcrypt

fpletz committed:
apparmor: fix build with libxcrypt

fpletz committed:
pam: enable libxcrypt by default

fpletz committed:
shadow: fix build with libxcrypt

fpletz committed:
systemd: fix build with libxcrypt

fpletz committed:
perl: fix build with libxcrypt

fpletz committed:
libxcrypt: fix build with musl

fpletz committed:
nixos/tests/shadow: new hashes support with libxcrypt

fpletz committed:
accountsservice: fix build with libxcrypt

fpletz committed:
openldap: fix build with libxcrypt

fpletz committed:
tdb: fix build with libxcrypt

fpletz committed:
llvmPackages_11.compiler-rt: fix build with libxcrypt

fpletz committed:
aprutil: fix build with libxcrypt

fpletz committed:
alpine: fix build with libxcrypt

fpletz committed:
talloc: fix build with libxcrypt

fpletz committed:
vsftpd: fix build with libxcrypt

fpletz committed:
apacheHttpd: fix build with libxcrypt

fpletz committed:
opensmtpd: fix build with libxcrypt

fpletz committed:
llvmPackage_14.compiler-rt: fix build with libxcrypt

fpletz commented:
@mweinelt I've continued your efforts and fixed most issues with obvious crypt users and most NixOS tests seem to be working including an expansion of the shadow test to check support of the new hashes. I've rebased and pushed my changes here https://github.com/NixOS/nixpkgs/compare/master...fpletz:nixpkgs:glibc-without-libcrypt. Is it okay if I push to your branch?

mweinelt commented:
Feel free!

mweinelt committed:
screen: fix build with libxcrypt

mweinelt committed:
pppd: fix build with libxcrypt

mweinelt committed:
pppd: inherit nixos test into passthru.tests

mweinelt commented on pkgs/tools/networking/ppp/default.nix:
```diff
@@ -48,6 +51,8 @@ stdenv.mkDerivation rec {
     "CC=${stdenv.cc.targetPrefix}cc"
   ];
 
+  NIX_LDFLAGS = "-lcrypt";
```
Without this the build would fail to link. Not sure if this is the proper way to workaround this.

```
ppp> cc -O2 -g -pipe -DHAVE_PATHS_H -DIPX_CHANGE -DHAVE_MMAP -pipe -I../include '-DDESTDIR="/nix/store/l0gbyq4xxzc053bswl9slcp7j0hnc3d6-ppp-2.4.9"' -DCHAPMS=1 -DMPPE=1 -DUSE_EAPTLS=1 -DHAS_SHADOW -I/nix/store/0yy468mw8yijm5p7y3gjk8wyf7qzh4g4-openssl-3.0.5-dev/include/openssl -DHAVE_MULTILINK -DUSE_TDB=1 -DPLUGIN -DPPP_FILTER -DINET6=1 -DMAXOCTETS  -Wl,-E -o pppd main.o magic.o fsm.o lcp.o ipcp.o upap.o chap-new.o md5.o ccp.o ecp.o auth.o options.o demand.o utils.o sys-linux.o ipxcp.o tty.o eap.o chap-md5.o session.o md4.o chap_ms.o sha1.o eap-tls.o pppcrypt.o multilink.o tdb.o spinlock.o ipv6cp.o eui64.o -lrt -lssl -lcrypto -ldl -lpcap
ppp> /nix/store/c5d4jq7mjsh8bkhy67k8mxsybj5r7p3x-binutils-2.38/bin/ld: auth.o: in function `check_passwd':
ppp> /build/source/pppd/auth.c:1515: undefined reference to `crypt'
ppp> /nix/store/c5d4jq7mjsh8bkhy67k8mxsybj5r7p3x-binutils-2.38/bin/ld: session.o: in function `session_start':
ppp> /build/source/pppd/session.c:346: undefined reference to `crypt'
ppp> collect2: error: ld returned 1 exit status
```

fpletz replied:
I've noticed that lots of build configuration scripts detect libcrypt very sloppily or not at all. I've also needed to add NIX_* at some places but sometimes a patch will also do. This was the weirdest one: https://github.com/NixOS/nixpkgs/pull/181764/commits/096f0e1d73c80fbc878c99d16a5445d6c95b1252

mweinelt committed:
open-vm-tools: fix build with libxcrypt

fpletz committed:
conserver: fix build with libxcrypt

fpletz committed:
dante: fix build with libxcrypt

fpletz committed:
hiawatha: fix build with libxcrypt

fpletz committed:
ircdHybrid: fix build with libxcrypt

mweinelt committed:
ldapvi: fix build with libxcrypt

mweinelt committed:
atheme: fix build with libxcrypt

fpletz committed:
libcli: fix build with libxcrypt

fpletz committed:
libguestfs: fix build with libxcrypt

fpletz committed:
leafnode: fix build with libxcrypt

mweinelt committed:
xorg.xdm: fix build with libxcrypt

mweinelt committed:
wmic-bin: fix build with libxcrypt

mweinelt committed:
libreswan: fix build with libxcrypt

fpletz committed:
lsh: fix build with libxcrypt

fpletz committed:
pgpool: fix build with libxcrypt

fpletz committed:
groonga: fix build with libxcrypt

fpletz committed:
mailutils: fix build with libxcrypt

fpletz committed:
popa3d: fix build with libxcrypt

mweinelt committed:
tengine: fix build with libxcrypt

mweinelt committed:
tcsh: fix build with libxcrypt

mweinelt committed:
swiProlog: fix build with libxcrypt

mweinelt committed:
sumo: fix build with libxcrypt

mweinelt committed:
toybox: fix build with libxcrypt

mweinelt committed:
gcc10: fix build with libxcrypt

mweinelt committed:
gcc12: fix build with libxcrypt

fpletz committed:
xrootd: fix build with libxcrypt

fpletz committed:
pounce: fix build with libxcrypt

fpletz committed:
uwsgi: fix build with libxcrypt

fpletz committed:
super: fix build with libxcrypt

fpletz committed:
policycoreutils: fix build with libxcrypt

fpletz committed:
pies: fix build with libxcrypt

fpletz committed:
partimage: fix build with libxcrypt

fpletz committed:
otpw: fix build with libxcrypt

fpletz committed:
zeroc-ice: fix build with libxcrypt

fpletz committed:
luaPackages.luaposix: fix build with libxcrypt

fpletz committed:
monit: fix build with libxcrypt

fpletz committed:
libfilezilla: fix build with libxcrypt

fpletz committed:
libsigrokdecode: fix build with libxcrypt

fpletz committed:
hylafaxplus: fix build with libxcrypt

fpletz committed:
dropbear: fix build with libxcrypt

fpletz committed:
kodi: fix build with libxcrypt

fpletz committed:
pure-ftpd: fix build with libxcrypt

fpletz committed:
cernlib: fix build with libxcrypt

fpletz committed:
bozohttpd: fix build with libxcrypt

fpletz committed:
sawfish: fix build with libxcrypt

fpletz committed:
qnial: fix build with libxcrypt

fpletz committed:
unicon-lang: fix build with libxcrypt

fpletz committed:
gvm-libs: fix build with libxcrypt

fpletz committed:
snis: fix build with libxcrypt

fpletz committed:
slock: fix build with libxcrypt

fpletz committed:
dico: fix build with libxcrypt

fpletz committed:
dcap: fix build with libxcrypt

fpletz committed:
haproxy: fix build with libxcrypt

fpletz committed:
srelay: fix build with libxcrypt

fpletz committed:
nntp-proxy: fix build with libxcrypt

fpletz committed:
shellhub-agent: fix build with libxcrypt

fpletz committed:
pam_mysql: fix build with libxcrypt

fpletz committed:
pam_pgsql: fix build with libxcrypt

fpletz committed:
sysvinit: fix build with libxcrypt

fpletz committed:
mokutil: fix build with libxcrypt

fpletz committed:
epic5: fix build with libxcrypt

fpletz committed:
kermit: fix build with libxcrypt

fpletz committed:
root5: fix build with libxcrypt

fpletz committed:
root: fix build with libxcrypt

fpletz committed:
pleroma: fix build with libxcrypt

fpletz committed:
sogo: fix build with libxcrypt

fpletz committed:
cde: fix build with libxcrypt

fpletz committed:
bftpd: fix build with libxcrypt

mweinelt committed:
llvmPackages_13.compiler-rt: fix build with libxcrypt

mweinelt committed:
llvmPackages_12.compiler-rt: fix build with libxcrypt

mweinelt committed:
llvmPackages_10.compiler-rt: fix build with libxcrypt

mweinelt committed:
llvmPackages_rocm.llvm: fix build with libxcrypt

mweinelt commented:
We pass the tested set :partying_face:

mweinelt committed:
haskellPackages.crypt-sha512: fix build with libxcrypt

and disable tests due to different ehaviour exhibited compared to
glibc when a low number of rounds is passed.

https://github.com/phadej/crypt-sha512/issues/13

yu-re-ka commented on pkgs/development/compilers/llvm/10/compiler-rt/default.nix:
```diff
@@ -25,6 +25,8 @@ stdenv.mkDerivation {
     "-DCOMPILER_RT_DEFAULT_TARGET_ONLY=ON"
     "-DCMAKE_C_COMPILER_TARGET=${stdenv.hostPlatform.config}"
     "-DCMAKE_ASM_COMPILER_TARGET=${stdenv.hostPlatform.config}"
+  ] ++ lib.optionals (haveLibc && !isMusl) [
```
Why is musl different from glibc here?

mweinelt replied:
https://gist.github.com/mweinelt/b1357859bf4a2f659a7b793db3ec7051

mweinelt commented:
- `pkgsStatic.libxcrypt` fails to configure, due to perl issues. `pkgsMusl.libxcrypt` works.

```
Can't locate Cwd.pm in @INC (you may need to install the Cwd module) (@INC contains: /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/site_perl/5.36.0/x86_64-linux /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/site_perl/5.36.0 /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/site_perl/5.36.0/x86_64-linux /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/site_perl/5.36.0/x86_64-linux /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/site_perl/5.36.0 /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/5.36.0/x86_64-linux /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/5.36.0) at /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/5.36.0/FindBin.pm line 84.
BEGIN failed--compilation aborted at /nix/store/fgnv3x7yaavvy1kjvrjbd19qk7a1cwr8-perl-static-x86_64-unknown-linux-musl-5.36.0/lib/perl5/5.36.0/FindBin.pm line 84.
Compilation failed in require at ./build-aux/scripts/expand-selected-hashes line 24.
BEGIN failed--compilation aborted at ./build-aux/scripts/expand-selected-hashes line 24.
configure: error: bad value 'all' for --enable-hashes

mweinelt committed:
freeswitch: fix build with libxcrypt

mweinelt committed:
haskellPackages.Unixutils: fix build with libxcrypt

mweinelt committed:
thttpd: fix build with libxcrypt

mweinelt committed:
haskellPackages.nano-cryptr: fix build with libxcrypt

mweinelt committed:
haskellPackages.xmonad-utils: fix build with libxcrypt

mweinelt committed:
ocserv: fix build with libxcrypt

mweinelt committed:
ladybird: fix build with libxcrypt

mweinelt committed:
cdesktopenv: fix build with libxcrypt

mweinelt committed:
wiringpi: fix build with libxcrypt

mweinelt committed:
swift: fix build with libxcrypt

mweinelt committed:
util-linux: pass libxcrypt, so sulogin gets built

Util-linux does not fail when crypt() isn't available. It just doesn't
build sulogin, which in turn breaks the initrd-builder for the
systemd-stage-1 tests.

mweinelt commented:
@ofborg eval

mweinelt commented:
This test failure of `python3.pkgs.twisted` looks suspicious.

```
python3.10-twisted> twisted.web.test.test_tap.ServiceTests.test_HTTPSFailureOnMissingSSL
python3.10-twisted> ===============================================================================
python3.10-twisted> [ERROR]
python3.10-twisted> Traceback (most recent call last):
python3.10-twisted>   File "/nix/store/kr37c537cx60b2j7km3rw4sx93p8wbwr-python3.10-twisted-22.8.0/lib/python3.10/site-packages/twisted/conch/test/test_checkers.py", line 104, in test_refuteCryptedPassword
python3.10-twisted>     crypted = crypt.crypt(password, password)
python3.10-twisted>   File "/nix/store/9cd6n41xfv06pv9n5yb558mc05h7lj5z-python3-3.10.7/lib/python3.10/crypt.py", line 82, in crypt
python3.10-twisted>     return _crypt.crypt(word, salt)
python3.10-twisted> builtins.OSError: [Errno 22] Invalid argument
python3.10-twisted> 
python3.10-twisted> twisted.conch.test.test_checkers.HelperTests.test_refuteCryptedPassword
python3.10-twisted> -------------------------------------------------------------------------------
python3.10-twisted> Ran 11905 tests in 362.704s
python3.10-twisted> 
python3.10-twisted> FAILED (skips=1694, errors=1, successes=10210)

mweinelt committed:
python3Packages.twisted: Disable failing crypt test

Fails due to a combination of libxcrypt and python upstream crypt
changes.

https://github.com/python/cpython/commit/0d3fe8ae4961bf551e7d5e42559e2ede1a08fd7c

mweinelt committed:
python3Packages.pillow{,-simd}: fix build with libxcrypt

mweinelt committed:
libxcrypt: Fix static build by using perl from buildPackages

mweinelt committed:
boost: fix build with enablePython by passing libxcrypt

mweinelt commented:
The comparison with trunk-combined looks pretty good. Most failed jobs are unrelated to this effort and most were fixed since the latest evaluation.

I aim to rebase and merge into staging after 2022/10/09 16:00 UTC (in ~17.5 hours) unless anything comes up.

maralorn commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -2531,6 +2531,34 @@ self: super: {
   # Restrictive upper bound on base.
   # Remove once version 1.* is released
   monad-bayes = doJailbreak super.monad-bayes;
+
+  crypt-sha512 = overrideCabal (drv: {
+    librarySystemDepends = [
+      pkgs.libxcrypt
+    ];
+    # Test failure after libxcrypt migration, reported upstrem at
+    # https://github.com/phadej/crypt-sha512/issues/13
+    doCheck = false;
+  }) super.crypt-sha512;
+
+  nano-cryptr = overrideCabal (drv: {
+    librarySystemDepends = [
+      pkgs.libxcrypt
+    ];
+  }) super.nano-cryptr;
+
+  Unixutils = overrideCabal (drv: {
+    librarySystemDepends = [
+      pkgs.libxcrypt
+    ];
+  }) super.Unixutils;
+
+  xmonad-utils = overrideCabal (drv: {
+    librarySystemDepends = [
+      pkgs.libxcrypt
+    ];
+  }) super.xmonad-utils;
+
```
Haskell changes look good in principle. (Can’t comment on the overall migration.)

Well, technically the `librarySystemDepends` override belong into `configuration-nix.nix` because they are permanent and will affect every version of those packages.

Only the `dontCheck` override belongs into `configuration-common.nix` because this is a temporary fix and will hopefully not be required by future versions of this package.

mweinelt replied:
Feel free to move these around, ideally by pushing fixup commits.

sternenseemann replied:
I think we can save ourselves quite a bit of work by adding a rule to `cabal2nix` that causes it to pick `libxcrypt` if it sees:

```cabal
   extra-libraries:  crypt
```

This could be as easy as changing [this line](https://github.com/NixOS/cabal2nix/blob/3a2abb2e24afe70da2e384f27ebc358ee9e0ebd3/src/Distribution/Nixpkgs/Haskell/FromCabal/Name.hs#L39).

sternenseemann replied:
@mweinelt What happens if we start injecting `libxcrypt` into Haskell builds on our regular branch already? Would glibc take precedence? Would it conflict?

mweinelt replied:
Honestly no idea.

vcunat replied:
I don't think it would conflict.  `ldd path/to/binary` should tell you if libxcrypt is used.

vcunat commented:
Tests can get weird sometimes
```
AssertionError: did not expect 'darwin' platform would have native support for 'bcrypt'
```
https://nix-cache.s3.amazonaws.com/log/1lhcrj4003daw01abrqshwlz3765qdyd-python3.10-passlib-1.7.4.drv

Artturin commented on pkgs/servers/x11/xorg/default.nix:
```diff
@@ -1992,7 +1992,7 @@ lib.makeScope newScope (self: with self; {
 
   # THIS IS A GENERATED FILE.  DO NOT EDIT!
-  xdm = callPackage ({ stdenv, pkg-config, fetchurl, libX11, libXau, libXaw, libXdmcp, libXext, libXft, libXinerama, libXmu, libXpm, xorgproto, libXrender, libXt }: stdenv.mkDerivation {
+  xdm = callPackage ({ stdenv, pkg-config, fetchurl, libX11, libXau, libXaw, libXdmcp, libXext, libXft, libXinerama, libXmu, libXpm, libxcrypt, xorgproto, libXrender, libXt }: stdenv.mkDerivation {
```
the changes to this file will be overriden once `./generate-expr-from-tarballs.pl tarballs.list` is run, moving this fix to overrides.nix in https://github.com/NixOS/nixpkgs/pull/199912
